src/main/java/com/kodilla/checkers/io/BoardGame.java

```diff
@@ -0,0 +1,316 @@
+                mouseY < startY + sideLength) // range check
+            return new BoardPosition( (int)((mouseX - startX) / unitLength),
+                    (int)((mouseY - startY) / unitLength ));
+        else return null;
+    }
+
+
+    public List<BoardPosition> getMoves(BoardPosition from) {
+        List<BoardPosition> result;
+
+        if (board.get(from).isCrown())
+            result = getStrikesCrown(from);
+        else result = getStrikes(from);
+
+        final int[] shifts = {-1, 1};
+        if (result.isEmpty() && !board.get(from).isEmpty()) {
+            if (board.get(from).isCrown())
+                for (int shiftX : shifts)
+                    for (int shiftY : shifts) {
+                        BoardPosition to = from.add(shiftX, shiftY);
+                        while (to.inBounds(board.side()) && board.get(to).isEmpty()) {
+                            result.add(to);
+                            to = to.add(shiftX, shiftY);
+                        }
+                    }
+            else for (int shift : shifts) {
+                BoardPosition move = from.add(new BoardPosition(shift,
+                        board.get(from).color() ? 1 : -1));
+                if (board.get(move) != null && board.get(move).isEmpty())
+                    result.add(new BoardPosition(move));
+            } }
+
+        for (BoardPosition pos : result)
+            pos.addToRoute(new BoardPosition(from));
+
+        return result;
+    }
+
+    private List<BoardPosition> getStrikes(BoardPosition from) {
+        Queue<BoardPosition> search = new LinkedList<>(); search.add(from);
+        List<BoardPosition> result = new ArrayList<>();
+        final int[] offsets = {-2, 2};
+
+        while (!search.isEmpty()) {
+            boolean finalPos = true;
+            for (int offX : offsets)
+                for (int offY : offsets) {
+                    BoardPosition to = new BoardPosition(search.peek().getX() + offX,
+                            search.peek().getY() + offY);
+                    to.setRoute(search.peek().getRoute());
+
+                    if (to.inBounds(board.side()) && board.get(to).isEmpty() &&
+                            !board.get(to.avg(search.peek())).isEmpty() &&
+                            board.get(from).color() !=
+                                    board.get(to.avg(search.peek())).color() &&
+                            !to.getRoute().contains(to.avg(search.peek()))) {
+                        to.addToRoute(new BoardPosition(to.avg(search.peek())));
+                        search.add(to);
+                        finalPos = false;
+                    }
+                }
+
+            if (finalPos && !search.peek().equals(from))
+                result.add(search.peek());
+            search.poll();
+        }
+
+        return filterShorter(result);
+    }
+
+    private List<BoardPosition> getStrikesCrown(BoardPosition from) {
```

Comment:
Szczerze powiedziawszy, szczegółowe zrozumienie tej metody wymaga heroizmu.
Gdy w pracy natkniemy się na taki kod i musimy go zmienić, podchodzimy następująco:
- ustalamy, co ta metoda właściwie ma robić, jaka jest jej odpowiedzialność,
- piszemy do niej testy, szczególnie uwzględniając wszystkie przypadki brzegowe, jakie jesteśmy w stanie wymyśleć,
- kawałek po kawałku refaktorujemy metodę, po każdej małej zmianie odpalamy testy.

W tym miejscu chciałbym jeszcze bardziej zaakcentować, jak testy są **ważne** dla długofalowego utrzymania i rozwoju aplikacji. Ich brak to w bardziej złożonych projektach po prostu samobójstwo.